File: lenses/applications/applications.go
package applications

import (
	"fmt"
	"os"
	"os/exec"
	"path/filepath"
	"strings"
	"syscall"
	"time"

	"github.com/indium114/spyglass/lens"

	"gopkg.in/yaml.v3"
)

type appConfig struct {
	Name        string `yaml:"name"`
	Icon        string `yaml:"icon"`
	Command     string `yaml:"command"`
	Description string `yaml:"description"`
	Context     []struct {
		Name    string `yaml:"name"`
		Command string `yaml:"command"`
	} `yaml:"context"`
}

type applicationsLens struct {
	apps []appConfig
}

func New() lens.Lens {
	l := &applicationsLens{}
	l.load()
	return l
}

func (a *applicationsLens) Name() string {
	return "Applications"
}

func (a *applicationsLens) load() {
	home, _ := os.UserHomeDir()
	dir := filepath.Join(home, ".config", "spyglass", "applications")
	files, _ := os.ReadDir(dir)

	for _, f := range files {
		if strings.HasSuffix(f.Name(), ".yaml") {
			data, _ := os.ReadFile(filepath.Join(dir, f.Name()))
			var cfg appConfig
			yaml.Unmarshal(data, &cfg)
			a.apps = append(a.apps, cfg)
		}
	}
}

func (a *applicationsLens) Search(query string) ([]lens.Entry, error) {
	var results []lens.Entry
	query = strings.ToLower(query)

	for _, app := range a.apps {
		if strings.Contains(strings.ToLower(app.Name), query) {
			results = append(results, lens.Entry{
				ID:          app.Name,
				Title:       app.Name,
				Icon:        app.Icon,
				Description: app.Description,
			})
		}
	}
	return results, nil
}

func (a *applicationsLens) Enter(entry lens.Entry) error {
	for _, app := range a.apps {
		if app.Name == entry.Title {
			cmd := exec.Command("sh", "-c", app.Command)
			cmd.Stdout = nil
			cmd.Stdin = nil
			cmd.Stderr = nil
			cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}

			if err := cmd.Start(); err != nil {
				return err
			}

			// ensure that the process has started before returning
			timeout := 10 * time.Second
			ticker := time.NewTicker(50 * time.Millisecond)
			defer ticker.Stop()
			deadline := time.After(timeout)

			for {
				select {
				case <-ticker.C:
					// check if process exists using signal 0
					if err := cmd.Process.Signal(syscall.Signal(0)); err == nil {
						return nil
					}
				case <-deadline:
					if err := cmd.Process.Signal(syscall.Signal(0)); err == nil {
						return nil
					}
					return fmt.Errorf("Application %q failed to start within %v", app.Name, timeout)
				}
			}
		}
	}
	return nil
}

func (a *applicationsLens) ContextActions(entry lens.Entry) []lens.Action {
	for _, app := range a.apps {
		if app.Name == entry.Title {
			var actions []lens.Action
			for _, c := range app.Context {
				command := c.Command
				actions = append(actions, lens.Action{
					Name: c.Name,
					Run: func(e lens.Entry) error {
						cmd := exec.Command("sh", "-c", command)
						cmd.Stdout = nil
						cmd.Stdin = nil
						cmd.Stderr = nil
						cmd.SysProcAttr = &syscall.SysProcAttr{Setsid: true}

						if err := cmd.Start(); err != nil {
							return err
						}

						// ensure that the process has started before returning
						timeout := 10 * time.Second
						ticker := time.NewTicker(50 * time.Millisecond)
						defer ticker.Stop()
						deadline := time.After(timeout)

						for {
							select {
							case <-ticker.C:
								// check if process exists using signal 0
								if err := cmd.Process.Signal(syscall.Signal(0)); err == nil {
									return nil
								}
							case <-deadline:
								if err := cmd.Process.Signal(syscall.Signal(0)); err == nil {
									return nil
								}
								return fmt.Errorf("Action %q failed to start within %v", c.Name, timeout)
							}
						}
					},
				})
			}
			return actions
		}
	}
	return nil
}

File: src/applications.rs
use crate::lens::{Entry, Lens};
use fuzzy_matcher::{FuzzyMatcher, skim::SkimMatcherV2};
use serde::{Deserialize, Serialize};
use std::{fs, io};

// helper
fn home() -> String {
    let dir = dirs::home_dir();
    dir.map(|p| p.to_string_lossy().into_owned())
        .unwrap_or_default()
}

#[derive(Serialize, Deserialize)]
struct AppContext {
    name: String,
    command: String,
}

#[derive(Serialize, Deserialize)]
pub struct AppConfig {
    name: String,
    icon: String,
    command: String,
    description: String,
    context: Vec<AppContext>,
}

pub struct ApplicationsLens {
    pub apps: Vec<AppConfig>,
}

impl ApplicationsLens {
    pub fn new() -> Self {
        Self { apps: Vec::new() }
    }

    pub fn load(&mut self) {
        let home = home();
        let dir = home + ".config/spyglass/applications";
        let files = fs::read_dir(dir).expect("failed to read ~/.config/spyglass/applications");

        for i in files {
            let i = i.unwrap();
            let config = fs::read_to_string(i.path())
                .ok()
                .and_then(|s| serde_yaml::from_str(&s).ok())
                .unwrap();

            self.apps.push(config);
        }
    }
}

impl Lens for ApplicationsLens {
    fn name(&self) -> &str {
        "Applications"
    }

    fn search(&self, query: &str) -> Result<Vec<crate::lens::Entry>, color_eyre::eyre::Error> {
        let mut results: Vec<Entry> = Vec::new();
        let matcher = SkimMatcherV2::default();
        for app in &*self.apps {
            if let Some(score) = matcher.fuzzy_match(&app.name, query) {
                results.push(Entry {
                    id: app.name.to_string(),
                    title: app.name.to_string(),
                    icon: app.icon.to_string(),
                    description: app.description.to_string(),
                });
            }
        }
        Ok(results)
    }

    fn enter(&self, entry: &crate::lens::Entry) -> Result<(), color_eyre::eyre::Error> {
        todo!()
    }

    fn context_actions(&self, entry: &crate::lens::Entry) -> Vec<crate::lens::Action> {
        Vec::new()
    }
}
